### PR TITLE
bedrock: check L1 origin changes when sequencing new blocks

### PR DESCRIPTION
**Pull request by protolambda**
@mslipper hit a bug where the sequencer is producing a block with bad data. To ensure we catch it when it happens again, so we can investigate the issue, we should defensively check the L1 origin changes.

### CODE REVIEW

File: op-node/rollup/driver/step.go [hunk 1]
```diff
@@ -37,9 +37,19 @@ func (d *outputImpl) createNewBlock(ctx context.Context, l2Head eth.L2BlockRef,
...
+				"head_origin", l2Head.L1Origin, "head_seq_nr", l2Head.SequenceNumber, "new_origin", l1Origin, "new_origin_parent", l1Origin.ParentID())
+			return l2Head, nil, fmt.Errorf("cannot create new block with L1 origin %s (parent %d) on top of L1 origin %s", l1Origin, l1Origin.ParentID(), l2Head.L1Origin)
+		}
 		l1Info, _, receipts, err = d.dl.Fetch(fetchCtx, l1Origin.Hash)
 		seqNumber = 0 // reset sequence number at the start of the epoch
 	} else {
+		if l2Head.L1Origin.Hash != l1Origin.Hash {
+			d.log.Error("SEQUENCING BUG: cannot create new block on top of l2Head with different origin at same height",
+				"head", l2Head, "head_origin", l2Head.L1Origin, "head_seq_nr", l2Head.SequenceNumber, "new_origin", l1Origin, "new_origin_parent", l1Origin.ParentID())
+			return l2Head, nil, fmt.Errorf("cannot create new block with L1 origin %s (parent %d) on top of L1 origin %s", l1Origin, l1Origin.ParentID(), l2Head.L1Origin)
```
Comment (trianglesphere) [inline op-node/rollup/driver/step.go:51]: How is %d working with `l1Origing.ParentID()`?

File: op-node/rollup/driver/step.go [hunk 2]
```diff
@@ -37,9 +37,19 @@ func (d *outputImpl) createNewBlock(ctx context.Context, l2Head eth.L2BlockRef,
 	// case we need to fetch all transaction receipts from the L1 origin block so we can scan for
 	// user deposits.
 	if l2Head.L1Origin.Number != l1Origin.Number {
+		if l1Origin.ParentHash != l2Head.L1Origin.Hash {
```
Comment (trianglesphere) [inline op-node/rollup/driver/step.go:40]: comparison is opposite above comparison

File: op-node/rollup/driver/step.go (hunk 1 above)
Comment (protolambda) [reply op-node/rollup/driver/step.go:51]: nice catch

File: op-node/rollup/driver/step.go (hunk 2 above)
Comment (protolambda) [reply op-node/rollup/driver/step.go:40]: I'm not sure if I understand your comment. If the l1 origin changes by number, and the parent of the new origin does not match the previous origin, then it is some inconsistent L1 chain reference and should thus error. That should be correct right?

Comment (protolambda) [reply op-node/rollup/driver/step.go:40]: Oh I see, you mean the order of the operands, not the equality check itself.

Comment (trianglesphere) [reply op-node/rollup/driver/step.go:40]: first comparison is `l2Head != l1Origin`, second is `l1Origin != l2Head`. The order of the arguments is different in each.

File: op-node/rollup/driver/step.go (hunk 1 above)
Comment (trianglesphere) [reply op-node/rollup/driver/step.go:51]: Also, I'm surprised a lint didn't catch this, I feel like they can look into format strings (or maybe that's the LSP).

File: op-node/rollup/driver/step.go (hunk 2 above)
Comment (protolambda) [reply op-node/rollup/driver/step.go]: Fixed (and fixed the fix, now all three comparisons match)

**Review by trianglesphere — APPROVED** [review]

**Pull request merged**